Sources/SourceKitLSP/Swift/MacroExpansion.swift

```diff
@@ -0,0 +1,122 @@
+//===----------------------------------------------------------------------===//
+//
+// This source file is part of the Swift.org open source project
+//
+// Copyright (c) 2014 - 2024 Apple Inc. and the Swift project authors
+// Licensed under Apache License v2.0 with Runtime Library Exception
+//
+// See https://swift.org/LICENSE.txt for license information
+// See https://swift.org/CONTRIBUTORS.txt for the list of Swift project authors
+//
+//===----------------------------------------------------------------------===//
+
+import Foundation
+import LSPLogging
+import LanguageServerProtocol
+import SourceKitD
+
+@_spi(Testing) public typealias MacroExpansionEdit = RefactoringEdit
+
+/// Detailed information about the result of a macro expansion operation.
+///
+/// Wraps the information returned by sourcekitd's `semantic_refactoring`
+/// request, such as the necessary macro expansion edits.
+struct MacroExpansion: RefactoringResponse {
+  /// The title of the refactoring action.
+  var title: String
+
+  /// The URI of the file where the macro is used
+  var uri: DocumentURI
+
+  /// The resulting array of `MacroExpansionEdit` of a semantic refactoring request
+  var edits: [MacroExpansionEdit]
+
+  init(title: String, uri: DocumentURI, refactoringEdits: [RefactoringEdit]) {
+    self.title = title
+    self.uri = uri
+    self.edits = refactoringEdits.compactMap { refactoringEdit in
+      if refactoringEdit.bufferName.isEmpty {
+        logger.error("Unable to retrieve some parts of the expansion")
+        return nil
+      }
+
+      return refactoringEdit
+    }
+  }
+}
+
+extension SwiftLanguageService {
+  /// Handles the `ExpandMacroCommand`.
+  ///
+  /// Makes request to sourcekitd and wraps the result into a `MacroExpansion`
+  /// and then makes `ShowDocumentRequest` to the client side for each
+  /// expansion to be displayed.
+  ///
+  /// - Parameters:
+  ///   - expandMacroCommand: The `ExpandMacroCommand` that triggered this
+  /// request.
+  ///
+  /// - Returns:
+  ///   - an `[MacroExpansionEdit]` with the necessary edits and buffer name as
+  /// a `LSPAny`
+  func expandMacro(
+    _ expandMacroCommand: ExpandMacroCommand
+  ) async throws -> LSPAny {
+    guard let sourceKitLSPServer else {
+      // `SourceKitLSPServer` has been destructed. We are tearing down the
+      // language server. Nothing left to do.
+      throw ResponseError.unknown("Connection to the editor closed")
+    }
+
+    guard let sourceFileURL = expandMacroCommand.textDocument.uri.fileURL else {
+      throw ResponseError.unknown("Given URL is not a file URL")
```

Comment:
Nit: LSP communicates in URIs instead of URLs

```suggestion
      throw ResponseError.unknown("Given URI is not a file URL")
```